src/MySqlConnector/MySqlClient/MySqlDataReader.cs

```diff
@@ -781,12 +830,32 @@ private enum State
 			NoMoreData,
 		}
 
+	    public async Task WaitDoneReadingAsync(CancellationToken cancellationToken=default(CancellationToken))
+	    {
+	        cancellationToken.ThrowIfCancellationRequested();
+	        await m_done_reading.WaitAsync(cancellationToken).ConfigureAwait(false);
+	        m_done_reading.Release();
+	    }
+
 		static readonly Task<bool> s_falseTask = Task.FromResult(false);
 		static readonly Task<bool> s_trueTask = Task.FromResult(true);
 
 		MySqlCommand m_command;
 		MySqlSession m_session;
-		State m_state;
+	    SemaphoreSlim m_done_reading = new SemaphoreSlim(1);
```

Comment:
Although `SemaphoreSlim` is lightweight, it seems wasteful to instantiate one of these for every single `MySqlDataReader` that's created (when most clients will not use it).

Prefer to initialise this in the constructor with every other field.